File: 백준/Silver/2805. 나무 자르기/나무 자르기.js
// /dev/stdin
const fs = require("fs");
let input = fs
  .readFileSync("/dev/stdin")
  .toString()
  .split("\n")
  .map((x) => x.replace("\r", ""));

const [N, M] = input[0].split(" ").map(Number);
const trees = input[1].split(" ").map(Number);
let answer = 0;

let left = 0;
let right = Math.max(...trees);

while (left <= right) {
  const mid = Math.floor((left + right) / 2);

  let high = 0;
  for (const tree of trees) {
    if (tree >= mid) {
      high += tree - mid;
    }
  }

  if (M <= high) {
    answer = mid;
    left = mid + 1;
  } else {
    right = mid - 1;
  }
}

console.log(answer);

File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
import sys


N, M = map(int, sys.stdin.readline().strip().split(' '))
trees = list(map(int, sys.stdin.readline().strip().split(' ')))

# print(trees)
def search(arr, target):
  left = 0
  right = max(arr)
  result = 0
  while left <= right:
    mid = (left + right) // 2
    # temp = [x for x in arr]
    # for i in range(len(temp)):
    #   if temp[i] - mid >= 0:
    #     temp[i] -= mid
    #   else:
    #     temp[i] = 0
    # print(temp)    
        
    # print(temp, mid, left, right)
    total = 0
    for x in arr:
      if x >= mid:
        total += (x - mid)

    if total >= target:
      result = mid
      left = mid + 1
    else:
      right = mid - 1

  return result

print(search(trees, M))


